build_runner/lib/src/server/hot_reload_client.dart

```diff
@@ -2,14 +2,85 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:async';
+import 'dart:convert';
 import 'dart:html';
+import 'dart:js';
 
+final _assetsDigestPath = r'$assetDigests';
 final _buildUpdatesProtocol = r'$livereload';
 
-main() {
+JsObject _dartLoader = context['\$dartLoader'];
+JsObject _urlToModuleId = _dartLoader['urlToModuleId'];
+
+List<String> get moduleUrls {
+  JsArray callMethod = context['Array']
```

Comment:
nit: Instead of creating a new array on the javascript side, you can just use `List.from` to make a copy on the Dart side.

Reply:
It doesn't work this way. `JsObject` isn't subtype of `Iterable`, and JS Iterator either

Reply:
If you don't use `dart:js` you won't get a `JsObject`.

Reply:
But there is no class in Dart I can map JS' Map or Iterator to. Or at least I failed to find one

Reply:
I don't understand the problem. What are you trying to do?

Reply:
How to get access to JS' Map's keys?
```
@JS(r'$dartLoader.urlToModuleId')
external Map urlToModuleId;
```
Fails when I do `urlToModuleId.keys` (no some method with dart-mangled name)

```
@JS
class JSIteratorItem {
 external bool get done;
 external dynamic get value;
}

@JS
class JsIterator {
  external JSIteratorItem next();
}

@JS(r'$dartLoader.urlToModuleId.keys')
external JsIterator urlToModuleIdKeys();
```
Something like this should work, but it too verbose, and I need to iterate it manually.

I wish I don't understand something and there is a better way.

Reply:
There is no such thing as a JS Map (well, other than ES6 `Map`, that is different).

Do you mean an object in JavaScript? If that is the case, you can write:

```dart
@JS(r'$dartLoader.urlToModuleId')
external Object get urlToModuleId;

@JS('Object.keys')
external List<dynamic> _jsObjectKeys(Object any);
List<String> jsObjectKeys(Object any) => List.from(_jsObjectKeys(any));

// Example use.
void main() {
  var keys = jsObjectKeys(urlToModuleId); // List<String>
}
```

Reply:
No, I mean exactly ES6 `Map`

Reply:
Ok, sorry about that. Then the following should work:

```dart
@JS(r'$dartLoader.urlToModuleId')
external JsMap get urlToModuleId;

@JS('Map')
abstract class JsMap {
  external Object keys();
}

@JS('Array.from')
external List<dynamic> jsArrayFrom(Object any);
```


... does that help?

Reply:
Yeah, thanks